core/src/main/java/com/opensymphony/xwork2/util/AnnotationUtils.java

```diff
@@ -139,19 +140,45 @@ public static String resolvePropertyName(Method method) {
      */
     public static <T extends Annotation> T findAnnotation(Class<?> clazz, Class<T> annotationClass) {
         T ann = clazz.getAnnotation(annotationClass);
-        while (ann == null && clazz != null) {
+        if (ann == null && clazz != Object.class) {
             ann = clazz.getAnnotation(annotationClass);
             if (ann == null) {
                 ann = clazz.getPackage().getAnnotation(annotationClass);
             }
-            if (ann == null) {
-                clazz = clazz.getSuperclass();
-                if (clazz != null) {
-                    ann = clazz.getAnnotation(annotationClass);
-                }
+            if (ann == null && clazz != Object.class) {
+                ann = findAnnotation(clazz.getSuperclass(), annotationClass);
             }
         }
 
         return ann;
     }
+
+    /**
+     * Returns the annotation on the given class or the package of the class. This searches up the
+     * class hierarchy and the package hierarchy for the closest match.
+     *
+     * @param <T>             class type
+     * @param clazz           The class to search for the annotation.
+     * @param annotationClass The Class of the annotation.
+     * @return The annotation or null.
+     */
+    public static <T extends Annotation> List<T> findAnnotations(Class<?> clazz, Class<T> annotationClass) {
+        List<T> anns = new ArrayList<>();
+
+        T ann = clazz.getAnnotation(annotationClass);
+        if (ann != null) {
+            anns.add(ann);
+        }
+
+        ann = clazz.getPackage().getAnnotation(annotationClass);
+        if (ann != null) {
+            anns.add(ann);
+        }
+
+        if (clazz.getSuperclass() != Object.class) {
```

Comment:
What do you feel with this instead:

```java
        ann = findAnnotations(clazz.getSuperclass(), annotationClass);
        if (ann != null) {
            anns.add(ann);
        }
```

Reply:
`findAnnotations` does not return `null` also I don't want to scan above `Object` as it doesn't make sense